Dockerfile.assisted-service-build

```diff
@@ -13,23 +10,36 @@ ENV PATH="$VIRTUAL_ENV/bin:$GOROOT/bin:$GOPATH/bin:$TOOLS:$PATH"
 COPY --from=quay.io/app-sre/golangci-lint:v1.37.1 /usr/bin/golangci-lint /usr/bin/golangci-lint
 COPY --from=quay.io/goswagger/swagger:v0.28.0 /usr/bin/swagger /usr/bin/goswagger
 COPY --from=quay.io/edge-infrastructure/swagger-codegen-cli:2.4.18 /opt/swagger-codegen-cli /opt/swagger-codegen-cli
-COPY --from=golang /usr/bin/gotestsum /usr/bin/make /usr/bin/
-COPY --from=golang /usr/local/go /usr/local/go
+COPY --from=registry.ci.openshift.org/openshift/release:golang-1.17 --chmod=774 /usr/bin/gotestsum /usr/bin/make /usr/bin/
+COPY --from=registry.ci.openshift.org/openshift/release:golang-1.17 --chmod=774 /usr/local/go /usr/local/go
 COPY --from=quay.io/openshift/origin-cli:latest /usr/bin/oc /usr/bin
 COPY --from=quay.io/operator-framework/upstream-opm-builder:v1.16.1 /bin/opm /bin
 COPY --from=k8s.gcr.io/kustomize/kustomize:v4.3.0 /app/kustomize /usr/bin/
 
-RUN dnf install -y 'dnf-command(config-manager)' && \
+RUN --mount=type=tmpfs,destination=/var/cache\
+    --mount=type=cache,target=/var/cache\
+    dnf install -y 'dnf-command(config-manager)' && \
     dnf config-manager --add-repo https://download.docker.com/linux/centos/docker-ce.repo && \
     dnf install -y \
-        openssl postgresql sqlite gcc genisoimage git docker-ce-cli libvirt-client libvirt-devel java && \
-    dnf clean all
+        openssl postgresql sqlite gcc genisoimage git docker-ce-cli libvirt-client libvirt-devel java
+
 
 WORKDIR /home/assisted-service
 RUN mkdir build $TOOLS --mode g+xw
 COPY ./hack/setup_env.sh ./dev-requirements.txt ./
-RUN ./setup_env.sh podman_remote && \
+RUN --mount=type=tmpfs,destination=/go/pkg/mod\
+    --mount=type=cache,target=/go/pkg/mod\
+    --mount=type=tmpfs,destination=/root/.cache\
+    --mount=type=cache,target=/root/.cache\
+    --mount=type=tmpfs,destination=/tmp/.cache\
+    --mount=type=cache,target=/tmp/.cache\
+    --mount=type=tmpfs,destination=/var/cache\
+    --mount=type=cache,target=/var/cache\
+    ./setup_env.sh podman_remote && \
     ./setup_env.sh assisted_service && \
-    dnf clean all && \
     chmod g+xw -R $GOPATH
 COPY ./data /tmp/data
+
+# merge layers
```

Comment:
This takes an empty image and copies over everything from the previous building process.
The advantage of this, is that we are effectively squashing all layers. Won't be very effective in development, as we will have dangling layers, but if this is used in CI or in PROD image it could bring benefits by cutting down the size of the image without going crazy with deduplicating every single duplicated file among layers

Reply:
How long this layer takes to be build? I think it's better to focus on developer experience as this image is mostly used along with skipper.